reps/2023-07-08-air-surface-syntax.md

```diff
@@ -0,0 +1,114 @@
+originally introduced for Ray AIR become unneccessary and the same functionality can nicely be achieved
+with the libraries themselves by making the libraries a little more interoperable.
+
+We have already implemented this strategy by replacing `BatchPredictor` with Ray Data native functionality
+(see https://github.com/ray-project/enhancements/blob/main/reps/2023-03-10-batch-inference.md and
+https://docs.ray.io/en/master/data/batch_inference.html) and by
+improving Ray Train's ingestion APIs
+(https://github.com/ray-project/enhancements/blob/main/reps/2023-03-15-train-api.md and
+https://docs.ray.io/en/master/ray-air/check-ingest.html).
+
+As a result of these changes, the `ray.air` namespace has become less and less relevant, and in this
+REP we propose to go all the way and remove it altogether in line with the Zen of Python
+```
+There should be one -- and preferably only one -- obvious way to do it.
+```
+This solves the confusions mentioned above and makes the Ray AIR APIs more coherent and focused around
+the cricital workloads (`ray.data` for batch inference, `ray.train` for training and `ray.serve` for serving).
+
+### Should this change be within `ray` or outside?
+
+main `ray` project. Changes are made to Ray Train, Tune and AIR.
+
+## Stewardship
+
+### Required Reviewers
+The proposal will be open to the public, but please suggest a few experienced Ray contributors in this technical domain whose comments will help this proposal. Ideally, the list should include Ray committers.
+
+@matthewdeng, @krfricke
+
+### Shepherd of the Proposal (should be a senior committer)
+To make the review process more productive, the owner of each proposal should identify a **shepherd** (should be a senior Ray committer). The shepherd is responsible for working with the owner and making sure the proposal is in good shape (with necessary information) before marking it as ready for broader review.
+
+@ericl
+
+## Details of the API changes
+
+Concretely, we replace the Ray AIR session with a training context to
+1. avoid the user confusion of what a `session` is (and not having to explain in the documentation) and
+2. bring the API in line with other Ray APIs like `get_runtime_context` as well as Ray Data's `DataContext`.
+
+The API changes are
+```
+from ray import air, train
+
+# Ray Train methods and classes:
+
+air.session.report               -> train.report
+air.session.get_dataset_shard    -> train.get_dataset_shard
+air.Checkpoint                   -> train.Checkpoint
+air.Result                       -> train.Result
+
+# Ray Train configurations:
+
+air.config.CheckpointConfig      -> train.CheckpointConfig
+air.config.FailureConfig         -> train.FailureConfig
+air.config.RunConfig             -> train.RunConfig
+air.config.ScalingConfig         -> train.ScalingConfig
+
+# Ray TrainContext methods:
+
+air.session.get_checkpoint       -> train.get_context().get_checkpoint
```

Comment:
IMO this should be top-level (`train.get_checkpoint`) c.f. `get_dataset_shard` and to avoid using the context for most workflows/examples

Reply:
So I think the reasoning in https://github.com/ray-project/ray/pull/36706#discussion_r1253792073 is still valid here and this probably should be in the context, but happy to revisit that with the checkpointing changes @ericl @justinvyu @matthewdeng :)

Reply:
To me `get_checkpoint` feels similar to both `get_dataset_shard` and `report`.

If I pass in a dataset, I call `get_dataset_shard`. If I pass in a checkpoint, I call `get_checkpoint`.

To save a checkpoint, I call `report`. To load a checkpoint, I call `get_checkpoint`.

With that said, if this is controversial we can always add an alias in the future.

Reply:
+1 to @ericl. 

I think the distinction can be made by separating configurations or Ray Train framework information vs. application information. I think this aligns with how context is used in other Ray libraries (like Ray DataContext).

So things like `experiement_name`, `trial_id`, `trial_resources` make sense under a context namespace.

However, informations that is useful for the application itself makes more sense as top level namespace. This includes `get_dataset_shard`, `get_checkpoint`, `local_rank`, `world_size`, etc. These concepts are not specific to Ray Train, but rather apply to ML training general. Unlike the previous category, this information is necessary for the actual ML application itself.

Reply:
Alright, I moved get_checkpoint into `train`.

I do think `local_rank` and `world_size` are infrequently enough accessed and close to a context that they should belong into the context. It is very analogous to `.get_node_id()` of `runtime_context` :)